### PR TITLE
Fix Gcp creds for workload identity

### LINKED ISSUE #394: Cannot determine Builder on GKE in Kubeflow 0.7 RC
/kind bug

**What steps did you take and what happened:**
Deployed Kubeflow 0.7 RC, running fairing/examples/prediction/xgboost-high-level-apis.ipynb, Failure in launching the training job on Kubeflow cluster with the error:
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-14-ed64d6eea92b> in <module>
      2 train_job = TrainJob(HousingServe, input_files=['ames_dataset/train.csv', "requirements.txt"],
      3                      docker_registry=DOCKER_REGISTRY,
----> 4                      backend=BackendClass(build_context_source=BuildContext))
      5 train_job.submit()

~/.local/lib/python3.6/site-packages/kubeflow/fairing/ml_tasks/tasks.py in __init__(self, entry_point, base_docker_image, docker_registry, input_files, backend, pod_spec_mutators)
     72                  input_files=None, backend=None, pod_spec_mutators=None):
     73         super().__init__(entry_point, base_docker_image, docker_registry,
---> 74                          input_files, backend, pod_spec_mutators)
     75 
     76     def submit(self):

~/.local/lib/python3.6/site-packages/kubeflow/fairing/ml_tasks/tasks.py in __init__(self, entry_point, base_docker_image, docker_registry, input_files, backend, pod_spec_mutators)
     58                                                  base_image=self.base_docker_image,
     59                                                  registry=self.docker_registry,
---> 60                                                  needs_deps_installation=needs_deps_installation)
     61         logger.warning("Using builder: {}".format(type(self.builder)))
     62 

~/.local/lib/python3.6/site-packages/kubeflow/fairing/backends/backends.py in get_builder(self, preprocessor, base_image, registry, needs_deps_installation, pod_spec_mutators)
    224                            "get change activated.")
    225             message = " ".join(msg)
--> 226             raise RuntimeError(message)
    227 
    228     def get_training_deployer(self, pod_spec_mutators=None):

RuntimeError: Not able to guess the right builder for this job!

**What did you expect to happen:**
Launch a training job on Kubeflow.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Fairing version: master @501adb8bd7ddb699a136b4f13268be67aa88a597
- Kubeflow version: 0.7-rc
- Kubernetes version: 1.16

**NOTE**: If you are using fairing from master, please provide us the git commit hash.

### PR DESCRIPTION
**Pull request by abhi-g**
**What this PR does / why we need it**:
Ensure appropriate credentials are used by Fairing jobs when running on GKE

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #394

**Special notes for your reviewer**:
The solutions uses the following approach:
1. When on GKE, check if user-gcp-sa credentials secret is available. If it is, use the secret.
1. If the secret is not available, assume Workload Identity is enabled. Use default-editor service account.
1. Modify pod_spec_template to include the service account name if secret is not mounted.

**Release note**:
```release-note
Incorporate workload identity into jobs launched via Fairing.
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/fairing/395)

### CODE REVIEW

File: kubeflow/fairing/builders/cluster/gcs_context.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def generate_pod_spec(self, image_name, push):  # pylint:disable=arguments-diffe
                 "--cache=true"]
         if not push:
             args.append("--no-push")
+
```
Comment (karthikv2k) [inline kubeflow/fairing/builders/cluster/gcs_context.py]: unrelated change?

File: kubeflow/fairing/cloud/gcp.py [hunk 2]
```diff
@@ -64,11 +64,17 @@ def add_gcp_credentials_if_exists(kube_manager, pod_spec, namespace):
...
                            .format(constants.GCP_CREDS_SECRET_NAME))
+            logger.warning("Trying workload identity service account: {}"\
+                           .format(constants.GCP_SERVICE_ACCOUNT_NAME))
+            pod_spec.service_account_name = constants.GCP_SERVICE_ACCOUNT_NAME
     except Exception as e:
         logger.warning("could not check for secret: {}".format(e))
 
 
 def add_gcp_credentials(kube_manager, pod_spec, namespace):
+    """Note: This method will be deprecated soon and will become unavailable by 1.0.
```
Comment (karthikv2k) [inline kubeflow/fairing/cloud/gcp.py]: use https://docs.python.org/3/library/exceptions.html#DeprecationWarning ? e.g. https://stackoverflow.com/questions/9008444/how-to-warn-about-class-name-deprecation

File: kubeflow/fairing/cloud/gcp.py [hunk 3]
```diff
@@ -64,11 +64,17 @@ def add_gcp_credentials_if_exists(kube_manager, pod_spec, namespace):
         else:
             logger.warning("Not able to find gcp credentials secret: {}"\
                            .format(constants.GCP_CREDS_SECRET_NAME))
+            logger.warning("Trying workload identity service account: {}"\
+                           .format(constants.GCP_SERVICE_ACCOUNT_NAME))
+            pod_spec.service_account_name = constants.GCP_SERVICE_ACCOUNT_NAME
```
Comment (karthikv2k) [inline kubeflow/fairing/cloud/gcp.py]: add a simple unit test for this function?

File: kubeflow/fairing/cloud/gcp.py (hunk 2 above)
Comment (abhi-g) [reply kubeflow/fairing/cloud/gcp.py]: Given that this function is only called by add_gcp_credentials_if_exists and is not directly used by any job configurator, not sure if all of that warning mechanism makes sense. Essentially, add_gcp_credentials_if_exists will become just adding the service_acccount_name to the pod_spec template.

File: kubeflow/fairing/cloud/gcp.py [hunk 4]
```diff
@@ -64,11 +64,17 @@ def add_gcp_credentials_if_exists(kube_manager, pod_spec, namespace):
         else:
             logger.warning("Not able to find gcp credentials secret: {}"\
                            .format(constants.GCP_CREDS_SECRET_NAME))
+            logger.warning("Trying workload identity service account: {}"\
```
[the remark this answers is not in the document]
Comment (abhi-g) [reply kubeflow/fairing/cloud/gcp.py]: done

File: kubeflow/fairing/builders/cluster/gcs_context.py (hunk 1 above)
Comment (abhi-g) [reply kubeflow/fairing/builders/cluster/gcs_context.py]: reverted
